sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Failure.java

```diff
@@ -0,0 +1,101 @@
+    return new AutoValue_Failure<>(exception, value);
+  }
+
+  public abstract Exception exception();
+
+  public abstract T value();
+
+  /**
+   * Internal class for collecting tuple tags associated with collections of {@link Exception}
+   * classes that should route to them. Also contains helper methods to simplify implementation of
+   * the {@code WithFailures} nested classes of {@link MapElements}, {@link FlatMapElements}, etc.
+   */
+  @AutoValue
+  abstract static class TaggedExceptionsList<T> implements Serializable {
+    abstract ImmutableList<TupleTag<Failure<T>>> tags();
+
+    abstract ImmutableList<List<Class<?>>> exceptionLists();
+
+    static <T> TaggedExceptionsList<T> empty() {
+      return new AutoValue_Failure_TaggedExceptionsList<>(ImmutableList.of(), ImmutableList.of());
+    }
+
+    /**
+     * Return a new {@link TaggedExceptionsList} that has all the tags and exceptions of this {@link
+     * TaggedExceptionsList} plus a new element representing the arguments passed in here.
+     */
+    TaggedExceptionsList<T> and(
+        TupleTag<Failure<T>> tag, Class<?> exceptionToCatch, Class<?>[] additionalExceptions) {
+      final ImmutableList<TupleTag<Failure<T>>> newTags =
+          ImmutableList.<TupleTag<Failure<T>>>builder().addAll(tags()).add(tag).build();
+      final ImmutableList<List<Class<?>>> newExceptionLists =
+          ImmutableList.<List<Class<?>>>builder()
+              .addAll(exceptionLists())
+              .add(
+                  ImmutableList.copyOf(ObjectArrays.concat(exceptionToCatch, additionalExceptions)))
+              .build();
+      return new AutoValue_Failure_TaggedExceptionsList<>(newTags, newExceptionLists);
+    }
+
+    /** Return the internal typed list of tags as an untyped {@link TupleTagList}. */
+    TupleTagList tupleTagList() {
+      TupleTagList l = TupleTagList.empty();
+      for (TupleTag<?> tag : tags()) {
+        l = l.and(tag);
+      }
+      return l;
+    }
+
+    /**
+     * Check the registered exception classes to see if the exception passed in here matches. If it
+     * does, wrap the exception and value together in a {@link Failure} and send to the output
+     * receiver. If not, rethrow so processing stops on the unexpected failure.
+     */
+    void outputOrRethrow(Exception e, T value, MultiOutputReceiver receiver) throws Exception {
+      for (int i = 0; i < tags().size(); i++) {
+        for (Class<?> cls : exceptionLists().get(i)) {
+          if (cls.isInstance(e)) {
+            receiver.get(tags().get(i)).output(Failure.of(e, value));
+            return;
+          }
+        }
+      }
+      throw e;
+    }
+
+    /**
+     * Set appropriate coders on all the failure collections in the given {@link PCollectionTuple}.
+     */
+    PCollectionTuple applyFailureCoders(PCollectionTuple pcs) {
+      final SerializableCoder<Failure<T>> failureCoder =
+          SerializableCoder.of(new TypeDescriptor<Failure<T>>() {});
```

Comment:
SerializableCoder also causes severe problems for streaming pipelines, because the encoding might change on pipeline update. Is it possible to write a custom FailureCoder?

Reply:
Also keep in mind that T is not guaranteed to be Serializable, so SerializableCoder may not even work here.

Reply:
>  keep in mind that T is not guaranteed to be Serializable, so SerializableCoder may not even work here

I had been assuming that T would have to implement `Serializable`.

I could add a FailureCoder, which I expect would be similar to `KvCoder`. But there's no default coder for `Exception`, so I need to figure out what to do there. It may be possible to use `AvroCoder` since `Exception` has a default constructor, but I'm not sure if that would fail for subclasses that don't have default constructors.

Reply:
> This is ugly and I regret saying it, but leaving it here for posterity

😆 . There are definitely some benefits to the approach of providing a failureHandler function and it would be more convenient than what's proposed here in some scenarios, so I find it to be a perfectly reasonable suggestion. 

Providing an output collection of failures will hopefully to more flexible, though, and better integrated with other parts of Beam.

Reply:
`SerializableCoder` has the nice property of being able to encode and recover the identity of the class itself. I've done some testing with AvroCoder and it looks like that would limit us to picking a specific type. So an `IOException` would get encoded/decoded as `Exception`.

We could make exception type a parameter, so `Failure<T>` would become `Failure<E, V>`, but some exception subclasses don't provide a no-argument constructor, and thus would not be decodable by AvroCoder.

Reply:
BTW an onFailure that takes an OutputReceiver is not a horrible option. It's also not mutually exclusive with the current approach.